gtfs/spec/en/reference.md

```diff
@@ -295,6 +305,68 @@ For examples that demonstrate how to specify a fare structure with [fare_rules.t
+| `spanning_limit` | Non-negative integer | **Conditionally Forbidden** | Defines how many consecutive legs the transfer rule may be applied to.<br><br>Valid options are:<br>`0` or empty - No limit.<br>`2` or more - Defines how many legs the transfer rule may span.<br><br>Conditionally Forbidden:<br>- **Forbidden** if `fare_transfer_rules.from_leg_group_id` does not equal `fare_transfer_rules.to_leg_group_id`.<br>- Optional if `fare_transfer_rules.from_leg_group_id` equals `fare_transfer_rules.to_leg_group_id`. |
+| `duration_limit` | Non-negative integer | Optional | Defines the duration limit of the transfer.<br><br>Must be expressed in integer increments of seconds.<br><br>If there is no duration limit, `fare_transfer_rules.duration_limit` must be empty. |
+| `duration_limit_type` | Enum | **Conditionally Required** | Defines the relative start and end of `fare_transfer_rules.duration_limit`.<br><br>Valid options are:<br>`0` - Between the departure fare validation of the first leg and the arrival fare validation of the last leg.<br>`1` - Between the departure fare validation of the first leg and the departure fare validation of the last leg.<br>`2` - Between the arrival fare validation of the first leg and the departure fare validation of the last leg.<br>`3` - Between the arrival fare validation of the first leg and the arrival fare validation of the last leg.<br><br>Conditionally Required:<br>- **Required** if `fare_transfer_rules.duration_limit` is defined.<br>- **Forbidden** if `fare_transfer_rules.duration_limit` is empty. |
+| `fare_transfer_type` | Enum | **Conditionally Required** | Indicates the cost processing method of leg-to-leg travel (i.e., sub-journey).<br><br>Valid options are:<br>`0` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of `fare_transfer_rules.amount`.<br>`1` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of `fare_transfer_rules.amount` plus the cost of `fare_leg_rules.amount` for `fare_transfer_rules.to_leg_group_id`.<br>`2` - The cost of the sub-journey is the cost of the leg with the most expensive `fare_leg_rules.amount` plus the cost in `fare_transfer_rules.amount`.<br>`3` - The cost of the two-leg sub-journey is the cost in `fare_transfer_rules.amount`.<br><br>Conditionally Required:<br>- **Required** if `fare_transfer_rules.amount` is defined.<br>- **Forbidden** otherwise. |
```

Comment:
This is not clear IMHO. As a starter this is the only place in the reference the phrase "sub-journey" is used, and I have no idea what it means. Please either clarify it here or at least add it to somewhere else in the fie. Further more it seems that the 1st two sentences are contradicting each other: "leg-to-leg travel (i.e., sub-journey)" hints that sub-journey is the transfer (from any leg N to leg N+1), while the next sentence "The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of fare_transfer_rules.amount" hints that sub-journey is any prefix of the full journey (leg1, leg2, ... legN, transfer).
The wording of the enum values also lack the mentioning of previous transfer conts: IMHO instead of "preceding leg(s)" it should be "preceding leg(s) and transfer(s)".
Value 3 mentions "two-leg sub-journey" which is confusing if we have more than 2 legs: (leg1,leg2)'s cost is "fare_transfer_rules.amount1", but then if the same(or other) rule applies to (leg2, leg3)=>amount2, then what's the cost of (leg1, leg2, leg3)? amount2? amount1+amount2?

Reply:
I agree that its confounding. Perhaps we can add some definitions in the "Term Definitions" section of the reference:

- **Leg** - Travel in which a rider boards and alights between a pair of subsequent locations along a trip. One leg, or multiple legs connected by transfers, may comprise a journey.
- **Sub-journey** - One leg, or multiple legs connected by transfers, that comprise a subset of a journey.
- **Journey** - Overall travel from origin to destination, including all legs and transfers in-between.

Thoughts on these definitions?

Reply:
Thanks for the thorough review!
> Further more it seems that the 1st two sentences are contradicting each other: "leg-to-leg travel (i.e., sub-journey)" hints that sub-journey is the transfer (from any leg N to leg N+1), while the next sentence "The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of fare_transfer_rules.amount" hints that sub-journey is any prefix of the full journey (leg1, leg2, ... legN, transfer).

I think along with the Term Definitions given above, we can update the first sentence of `fare_transfer_rules.fare_transfer_type` to simply:
> Indicates the cost processing method of a sub-journey.

These corrections would account for the fact that (i) a sub-journey IS a prefix of the full journey, and (ii) a single leg can be considered a sub-journey. So you are correct that "leg-to-leg" travel is not correct in the current definition.

>The wording of the enum values also lack the mentioning of previous transfer conts: IMHO instead of "preceding leg(s)" it should be "preceding leg(s) and transfer(s)".

Good point. I'm happy to add "and transfer(s)" to the enum descriptions as this is indeed the intended meaning.

> Value 3 mentions "two-leg sub-journey" which is confusing if we have more than 2 legs: (leg1,leg2)'s cost is "fare_transfer_rules.amount1", but then if the same(or other) rule applies to (leg2, leg3)=>amount2, then what's the cost of (leg1, leg2, leg3)? amount2? amount1+amount2?

The idea is that enum 3 is processing a cost interaction between two legs. The case of (leg1, leg2) yields the cost of `fare_transfer_rules.amount`. This becomes "embodied" as the cost of the "preceding leg(s) and transfer(s)". So when we get to (leg2, leg3), we are actually dealing with ((leg1,leg2), leg3) in which `fare_transfer_rules.amount` of (leg1, leg2) -> (leg 3) becomes the cost. Enum 3 is certainly a bit of a speculative feature, but its intended to be used for special cases where transferring to a certain leg has a cost that overrides any connecting travel beforehand. So I don't anticipate that a case where `fare_transfer_type=3` at every step between (leg1, leg2, leg3) would be common. 

Perhaps we can remove the confounding part on "two-leg sub-journey" and reword to:
> The cost of the sub-journey from the preceding leg(s) and transfer(s) to the transferred leg is the cost in `fare_transfer_rules.amount`.

Does that clarify the behaviour?

Reply:
1. leg vs sub-journey vs journey: even these definitions aren't clear enough to me. For example: does a sub-journey have to be contiguous? Or is leg1,leg4,leg7 also a sub-journey? To me the mathematical meaning of sub-set would allow it, and the "every day usage" might or might not allow it, however as someone who implemented a fare calculator I can say that it is a very important optimisation feature (i.e: if there's a bus ticket that allows you unlimited transfers within 2 hours then a "cheap" (well in this case expensive ;) fare calculator might do this: leg1:bus, leg2:tram, leg3:bus, leg4:bus so you pay 1 bus ticket on leg1, 1 tram ticket on leg2 and another bus ticket on leg3-4. however if we enable (as I interpret) non-contiguous sub-journeys then my fare calculator would spare you the 2nd bus ticket because if the whole journey takes only 90 minutes then the bus ticket from the 1st leg is sill valid for the other 2 bus legs)

2. I disagree with your comment: "a sub-journey IS a prefix of the full journey" This makes no sense at all (not mathematically and not even in real-world talk) If there are 4 legs and the first two legs are tram, and the last two legs are bus, then why leg1 and 2 can be a sub-journey but leg 3 and 4 can't? In my opinion leg3-4 should be a sub-journey as well.

If we assume that a sub-journey has to be a prefix of the journey then many things will be more complicated to represent in the files (while it may be easier to implement the fare calculator) or it is possible that the calculated total price of a journey will be higher than it really should be.

3. The biggest problem with how both the current wording and your explanation in the comment is that they both seem to assume that the fares are calculated this way (left to right):
leg1, leg1-to-leg2-transfer, leg2, leg2-to-leg3-transfer, leg3
While this is easy to implement it is IMHO not always the case (see above). The biggest issue with this (even if I go with your definition when a sub-journey is a prefix) is that this doesn't seem to be how real-life fares work. Whit what you propose in every point in the above left-to-right chain you can either add to the previous sum or throw it all away and swap it to a new amount. I can see how this work for 1 or 2 legs, but it gets hard to see how can this work for more than 2 legs.

4. even the file name and the enum values don't make much sense. If we're in the file called fare_transfer_rules.txt (BTW I hate txt, these files should be called csv) then why do some of the fare_transfer_type values try to enforce things that are "out of scope" aka override fares for "previous" legs.

5. Actually when we started our extension of the V1 fares we also thought about having 2 (or more) types of rules: leg and transfer. But at the end we ended up having 1 type of rule with almost all fields being optional, and this gives us great flexibility. Just to give a few examples: we have from_route_id, to_route_id. Both are optional, but look what we can achieve with them: if you only have to_route_id it is similar to what you would describe in the fare_leg_rules. But if we have both from_route_id and to_route_id then it kind of becomes a fare_transfer_rule.

We also can have zone_count or stop_count in our rules. These are neither leg nor transfer type, because they are defined by the counts on a sub-journey.

One feature we currently have is that we can have a "from_fare_id" in our rules. This allows us to have things that are similar to real world rules, like: if you have a valid $3 bus ticket in your wallet already from ANY of the previous legs, then you can use it on this leg. Or you can upgrade it to a $5 ticket. (One more reason to stick with fare_ids... You'll never be able to add future extensions like this, because you might have multiple fares (ticket types) that "accidentally" both cost $3, but but you can only upgrade the fare_id:1 ($3) to fare_id:7 ($5), because both of these belong to the same agency and are bus tickets, but you can't upgrade fare_id:18 ($3) because it belongs to another agency or only allows you to travel on the subway and without transfers....)

And because we have 1 huge rules table we can also combine the leg-type, the transfer-type and the sub-journey-type constraints within one rule.

Reply:
@flocsy 
- Leg vs sub-journey vs journey: The understanding is that a sub-journey is contiguous, and not just any subset. This is indicated in the definition that the legs must be "connected by transfers", but I will add the word "contiguous" to make this more clear.
- "Prefix": Good catch, that was certainly not a valid definition of a "sub-journey". 
- Your extended fares model: Do you have documentation published somewhere that we can reference? I'm interested in seeing this in detail.

<hr>

### Fare transfer processing

I took up the exercise of defining `fare_transfer_type` processing and interactions to get more precise about how they should be used and have a clear consensus on this.

Taking the current definitions, this is how a 2-leg transfer would be processed (variables represent the cost of the leg or transfer):
![2-leg transfer](https://user-images.githubusercontent.com/62957380/144116057-d5cf7ef0-1a67-4e65-922c-10140e410459.png)

| `fare_transfer_type` | Processing A > B |
| ----- | ----- |
| `0` | A + AB |
| `1` | A + AB + B |
| `2` | MAX(A, B) + AB |
| `3` | AB |

<hr>
Now where things get interesting, 3-leg transfer interactions:

![3-leg-transfer](https://user-images.githubusercontent.com/62957380/144116095-8a0aedae-5235-4782-a58f-e0b3644dd5f2.png)

Piecewise processing: 
- `[]>` and `>[]` indicates that the incoming transfer process result can be "plugged" into an outgoing transfer process.
- `+` indicates that the outgoing transfer process result is added to the incoming transfer process result.

| `fare_transfer_type` | Processing A > B | Processing B > C |
| ----- | ----- | ----- | 
| `0` | [A + AB]> | >[B] + BC |
| `1` | [A + AB + B]> | >[B] + BC + C |
| `2` | [MAX(A, B) + AB]> | + MAX(B, C) + BC |
| `3` | [AB]> | + BC |

Looking at how these would interact in different arrangements, I think all the cases are straightforward, except for 2>2. I see two reasonable options, delimited by "?or" in the table below. 

1. Process each leg pair separately, and then choose the max (leg+transfer) price.
2. Process each leg pair piecewise and add them together.

I think both of these behaviours are equally theoretical to how an agency might structure their fare policy, and as I've mentioned before, chaining `fare_transfer_type=2` seems to be also be an unlikely scenario by a data producer. At any rate, I prefer option 2 as it is more intuitive with the "plug" rules defined in the table above. Meanwhile, option 1 is a fairly arbitrary abstraction on the rules governing a 2-leg transfer (abstracting the same logic to a 3-leg transfer).

| B > C<hr>A > B | `0` | `1` | `2` | `3` | 
| ----- | ----- | ----- | ----- | ----- |
| `0` | A + AB + BC | A + AB + BC + C | A + AB + MAX(B, C) + BC | A + AB + BC | 
| `1` | A + AB + B + BC | A + AB + B + BC + C | A + AB + B + MAX(B, C) + BC | A + AB + B + BC | 
| `2` | MAX(A, B) + AB + BC | MAX(A, B) + AB + BC + C | MAX((MAX(A, B) + AB), (MAX(B, C) + BC))<br><br>?or<br><br>MAX(A, B) + AB + MAX(B, C) + BC | MAX(A, B) + AB + BC | 
| `3` | AB + BC | AB + BC + C | AB + MAX (B, C) + BC | AB + BC | 

\*Leg A>B is on the row header<br>\*Leg B>C is on the column header

Thoughts on these defined rules and `fare_transfer_type=2`?

Reply:
I think that it's better to have an additional fare_transfer_type to theoretically allow both ways, even if it's unlikely that a producer will use it. If you don't agree, then the description should be changed to make it crystal clear which is the intended usage.

I like your tables, and maybe it would be a good idea to add them to the spec (once we agree here on how they should be)

As for the notations, where you plug in the results of previous legs/transfers but you still use [B] on the right side, I think it would be clearer if instead of [B] it would be something like S (for Sum). ie:
0	[A + AB]	S + BC
and because of this I would also change the textual description to describe the steps this way, so that it'll be cleares when you look at both the text and the tables.

Reply:
I committed these changes here: 6e2abbc. Feel free to review.